providers/implementations/rands/fips_crng_test.c

```diff
@@ -0,0 +1,428 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * Implementation of SP 800-90B section 4.4 Approved Continuous Health Tests.
+ */
+
+#include <string.h>
+#include <openssl/evp.h>
+#include <openssl/core_dispatch.h>
+#include <openssl/params.h>
+#include <openssl/self_test.h>
+#include <openssl/proverr.h>
+#include "prov/providercommon.h"
+#include "prov/provider_ctx.h"
+#include "prov/implementations.h"
+#include "internal/cryptlib.h"
+#include "crypto/rand_pool.h"
+#include "drbg_local.h"
+#include "prov/seeding.h"
+#include "crypto/context.h"
+
+static OSSL_FUNC_rand_newctx_fn crng_test_new;
+static OSSL_FUNC_rand_freectx_fn crng_test_free;
+static OSSL_FUNC_rand_instantiate_fn crng_test_instantiate;
+static OSSL_FUNC_rand_uninstantiate_fn crng_test_uninstantiate;
+static OSSL_FUNC_rand_generate_fn crng_test_generate;
+static OSSL_FUNC_rand_reseed_fn crng_test_reseed;
+static OSSL_FUNC_rand_gettable_ctx_params_fn crng_test_gettable_ctx_params;
+static OSSL_FUNC_rand_get_ctx_params_fn crng_test_get_ctx_params;
+static OSSL_FUNC_rand_verify_zeroization_fn crng_test_verify_zeroization;
+static OSSL_FUNC_rand_enable_locking_fn crng_test_enable_locking;
+static OSSL_FUNC_rand_lock_fn crng_test_lock;
+static OSSL_FUNC_rand_unlock_fn crng_test_unlock;
+static OSSL_FUNC_rand_get_seed_fn crng_test_get_seed;
+static OSSL_FUNC_rand_clear_seed_fn crng_test_clear_seed;
+
+#ifndef ENTROPY_H
+# define ENTROPY_H 6    /* default to six bits per byte of entropy */
```

Comment:
This seems like something that would  need to be settable.

Reply:
My intention was to set this via `-DENTROPY_H=xxx` on the configure line.  It's not something we can set without knowing the particular entropy source being used and that will require us to put the source through validation.  This isn't something the project is likely to do because of the number of platforms supported.

The default of _6_ is fairly conservative for the sources we support.  The OS source should be 8 bits per byte e.g.

Reply:
I guess we could ask upstream when it quality is like and use that value and default to something if it doesn't answer.

Maybe worth delaying until we can have an entropy source inside the FIPS provider.

Reply:
It could be.  I'm not sure how exactly we quality of a seed source should be implemented.  I was planning on deferring that until such time as we have a seed source inside the FIPS boundary.

The reason for choosing 6 is based on the best result from the non-IID stream of entropy assessment.  In actual use, the should be specified specifically per seed source & platform.